### PR TITLE
Add structType support

### PR DESCRIPTION
**Pull request by mariantalla**
Similarly to https://github.com/apelisse/controller-tools/pull/1, this PR adds a `+structType` marker, that maps to `x-kubernetes-map-type` in kube-openapi.

cc @apelisse

### CODE REVIEW

**Comment by mariantalla** [comment]
(~Oops sorry, I'll rebase~ edit: done)

Putting WIP for now because I see test failures due to this validation:

```
if schema.Type != "object" {
    return fmt.Errorf("must apply structType to an object")
}
```

Test failures show that `schema.Type` is empty, which is a bit surprising 🤔 Looking into this next.

File: pkg/crd/markers/topology.go [hunk 1]
```diff
@@ -29,6 +29,7 @@ var TopologyMarkers = []*definitionWithHelp{
 	must(markers.MakeDefinition("listType", markers.DescribesField, ListType(""))),
 	must(markers.MakeDefinition("listMapKey", markers.DescribesField, ListMapKey(""))),
 	must(markers.MakeDefinition("mapType", markers.DescribesField, MapType(""))),
+	must(markers.MakeDefinition("structType", markers.DescribesField, StructType(""))),
```
Comment (apelisse) [inline pkg/crd/markers/topology.go:32]: Keeping them ordered, sweet

Comment (apelisse) [reply pkg/crd/markers/topology.go:32]: listType and listMapKey are not ordered :-/ Not your fault ...

File: pkg/crd/markers/topology.go [hunk 2]
```diff
@@ -110,4 +124,21 @@ func (m MapType) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
...
+		return fmt.Errorf(`StructType must be either "granular" or "atomic"`)
+	}
+
+	p := string(s)
+	schema.XMapType = &p
+
+	return nil
+}
+
+func (s StructType) ApplyFirst() {}
```
Comment (apelisse) [inline pkg/crd/markers/topology.go]: why is this needed?

File: pkg/crd/markers/topology.go [hunk 3]
```diff
@@ -110,4 +124,21 @@ func (m MapType) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
 	return nil
 }
 
-func (l MapType) ApplyFirst() {}
+func (m MapType) ApplyFirst() {}
+
+func (s StructType) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
+	if schema.Type != "object" {
```
Comment (apelisse) [inline pkg/crd/markers/topology.go]: I suspect "object" is the default and is thus not required.

**Comment by apelisse** [comment]
One minor nit, but looks good otherwise.

File: pkg/crd/markers/topology.go (hunk 1 above)
Comment (mariantalla) [reply pkg/crd/markers/topology.go:32]: > listType and listMapKey are not ordered

Shall I change that as part of this PR? It's a tiny bit of scope creep, but especially if we'll squash commits as part of https://github.com/kubernetes-sigs/controller-tools/pull/347, it's probably not a huge deal.

File: pkg/crd/markers/topology.go (hunk 2 above)
Comment (mariantalla) [reply pkg/crd/markers/topology.go]: Mainly driven by this: https://github.com/kubernetes-sigs/controller-tools/blob/4752ed2de7d2fc1b6b18398bf26cf2ce6b53cd94/pkg/crd/markers/doc.go#L27-L29. If I got it right, it suggests that "top-level" types (e.g. lists, as opposed to list items) can implement this method to ensure that their schemas will be applied before types that don't implement it.

The comment also says it's `discouraged` and `a hack` so if all this is not quite right, I'm happy to remove it. No tests fail if I remove `ApplyFirst` for StructType, for what it's worth 🤔

File: pkg/crd/markers/topology.go (hunk 3 above)
Comment (mariantalla) [reply pkg/crd/markers/topology.go]: Aha, makes sense. I [went for](https://github.com/kubernetes-sigs/controller-tools/blob/4752ed2de7d2fc1b6b18398bf26cf2ce6b53cd94/pkg/crd/markers/doc.go#L27-L29) `"object" || ""` just to be on the strict side. Let me know what you think, happy to remove the check completely if we think it doesn't add much.

File: pkg/crd/markers/topology.go (hunk 1 above)
Comment (apelisse) [reply pkg/crd/markers/topology.go:32]: Yeah, I'll re-order when I rebase/squash! :-)

File: pkg/crd/markers/topology.go (hunk 2 above)
Comment (apelisse) [reply pkg/crd/markers/topology.go]: I would only use it if there is an actual dependency between two types (like we do for ListType). But I don't think that's the case here.

Comment (mariantalla) [reply pkg/crd/markers/topology.go]: Removed in https://github.com/apelisse/controller-tools/pull/2/commits/5d66dadae91d4ac76bfc72472903f5ab8826dc87

**Comment by apelisse** [comment]
Thanks!
